src/main/java/com/google/devtools/build/lib/bazel/bzlmod/BzlmodRepoRuleCreator.java

```diff
@@ -93,4 +99,51 @@ public static Rule createRule(
     }
     return rule;
   }
+
+  public static void validateLabelAttrs(Rule rule, ModuleExtensionId owningExtension)
+      throws EvalException {
+    for (var attribute : rule.getRuleClassObject().getAttributes()) {
+      Object value = rule.getAttr(attribute.getName());
+      Collection<?> toValidate =
+          switch (value) {
+            case List<?> list -> list;
+            case Map<?, ?> map -> map.keySet();
+            case null, default -> singletonList(value);
+          };
+      for (var item : toValidate) {
+        if (!(item instanceof Label label)) {
+          continue;
+        }
+        if (label.getRepository().isVisible()) {
+          continue;
+        }
+        String unprefixedRepoName = label.getName().substring(label.getName().lastIndexOf('~') + 1);
+        RepositoryName owningModuleRepoName = owningExtension.getBzlFileLabel().getRepository();
+        String owningModule;
+        if (owningModuleRepoName.isMain()) {
+          owningModule = "root module";
+        } else {
+          owningModule =
+              String.format(
+                  "module '%s'",
+                  owningModuleRepoName
+                      .getName()
+                      .substring(0, owningModuleRepoName.getName().indexOf('~')));
+        }
+        throw Starlark.errorf(
+            "no repository visible as '@%s', but referenced by label '@%s//%s:%s' in attribute %s"
+                + " of %s '%s'. Only repositories visible to the %s can be referenced"
```

Comment:
s/`the %s`/`the module '%s'`

Reply:
That's in `owningModule` since it has to be slightly different for the `root module`.

Reply:
oh hang on, I just understood this. `owningModule` is a string. ignore me then!